torch/csrc/jit/tensorexpr/cuda_half_support.h

```diff
@@ -50,8 +74,16 @@ class CudaHalfChecker : public IRMutator {
     // just don't allow half casts we didn't insert.
     if (v->dtype().scalar_type() == ScalarType::Half) {
       if (inserted_half_casts_.count(v) < 1) {
-        // TODO: discards lanes.
-        return new Cast(kFloat, child);
+        return child;
+      }
+    }
+
+    // Remove Half(Float()) and friends.
+    const Cast* cast_child = dynamic_cast<const Cast*>(child);
+    if (cast_child) {
+      if (v->dtype().is_floating_point() &&
```

Comment:
I agree with your earlier comment 
> Normally they'd be in the simplifier, but I'm not confident we always want to eliminate Float(Half(X)) since it does have an affect (reducing precision).

What about specifically the case of casting from a type to a new type and back again ? I think that can generally be removed without reducing precision.

Reply:
I don't see how, consider: `Int(Char(Int(X)))` when X is FP. Both the cast to Int and the cast to Char have effects here.

I'm not opposed to the idea of a general casting simplification but I think it's somewhat complex and don't want to roll it into this fix.

Reply:
Ah true, i guess it's just lower precision -> higher precision -> lower precision where both types are floating point or both are integers